File: DownloadTileCacheSample/DownloadTileCacheSample/ViewController.m
// Copyright 2013 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

#import "ViewController.h"
#import "BackgroundHelper.h"
#import "SVProgressHUD.h"
#import "MessageHelper.h"

@interface ViewController ()
@end

@implementation ViewController

// in iOS7 this gets called and hides the status bar so the view does not go under the top iPhone status bar
- (BOOL)prefersStatusBarHidden
{
    return YES;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    
    //You can change this to any other service on tiledbasemaps.arcgis.com if you have an ArcGIS for Organizations subscription
    NSString* tileServiceURL = @"http://sampleserver6.arcgisonline.com/arcgis/rest/services/World_Street_Map/MapServer";
    
    NSURL *tiledUrl = [[NSURL alloc] initWithString:tileServiceURL];
    self.tiledLayer = [[AGSTiledMapServiceLayer alloc] initWithURL:tiledUrl];
    self.tiledLayer.delegate  = self;
    
    [self.mapView addMapLayer:self.tiledLayer withName:@"World Street Map"];
    
       
    // Init the tile cache task
    if ( self.tileCacheTask == nil) {
        NSURL *tiledUrl = [[NSURL alloc] initWithString:tileServiceURL];
        self.tileCacheTask = [[AGSExportTileCacheTask alloc] initWithURL:tiledUrl];
    }

    self.scaleLabel.numberOfLines = 0;
    

}



- (void)layer:(AGSLayer *)layer didFailToLoadWithError:(NSError *)error{
    [[[UIAlertView alloc]initWithTitle:@"Error" message:[error localizedDescription] delegate:Nil cancelButtonTitle:nil otherButtonTitles:nil, nil] show];
}

- (void)layerDidLoad:(AGSLayer *)layer{
    if (layer == self.tiledLayer) {
        self.scaleStepper.value = 0;
        self.scaleStepper.minimumValue = 0;
        self.scaleStepper.maximumValue = self.tiledLayer.tileInfo.lods.count;
        [self.mapView addObserver:self forKeyPath:@"mapScale" options:NSKeyValueObservingOptionNew context:NULL];
    }
}

- (void) observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context{
    self.estimateLabel.text = @"";
    self.scaleLabel.text = @"";
    self.lodLabel.text = @"";
    self.estimateButton.enabled = NO;
    self.downloadButton.enabled = NO;

    
    NSInteger index = [self.tiledLayer.mapServiceInfo.tileInfo.lods indexOfObject:self.tiledLayer.currentLOD];
    self.scaleStepper.maximumValue = self.tiledLayer.tileInfo.lods.count - index;
    self.scaleStepper.minimumValue = 0;
    self.scaleStepper.value = 0;
}

- (AGSLOD*) bestFitLODForEnvelope:(AGSEnvelope*)env{
    CGRect screenRect = [self.mapView toScreenRect:env];
    double impliedResolution = env.width /    screenRect.size.width;
    
    AGSMapServiceInfo *mapServiceInfo = self.tiledLayer.mapServiceInfo;
    
    for ( int i=0; i < mapServiceInfo.tileInfo.lods.count; i++ ) {
        AGSLOD * tempLod = [mapServiceInfo.tileInfo.lods objectAtIndex:i];
        if(impliedResolution > tempLod.resolution)
            return tempLod;
    }
    return [mapServiceInfo.tileInfo.lods lastObject];

}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)changeLevels:(id)sender
{
    self.estimateButton.enabled = YES;
    self.downloadButton.enabled = YES;
    self.scaleStepper.minimumValue = 1;
    
    AGSLOD * lod = [self.tiledLayer.mapServiceInfo.tileInfo.lods objectAtIndex:self.scaleStepper.value];
    self.lodLabel.text = [[NSNumber numberWithInt:self.scaleStepper.value] stringValue];
    self.scaleLabel.text = [NSString stringWithFormat:@"1:%@\n  to\n1:%@", [NSNumberFormatter localizedStringFromNumber:[NSNumber numberWithInt:self.tiledLayer.currentLOD.scale] numberStyle:NSNumberFormatterDecimalStyle],[NSNumberFormatter localizedStringFromNumber:[NSNumber numberWithInt:lod.scale] numberStyle:NSNumberFormatterDecimalStyle]];
}

- (IBAction)estimateAction:(id)sender
{
    
    NSArray *arrayLods = [self generateLods];
    NSLog(@"LODs %@", arrayLods);
    
    AGSEnvelope *extent = [self.mapView visibleAreaEnvelope];
    
    
    AGSExportTileCacheParams *params = [[AGSExportTileCacheParams alloc] initWithLevelsOfDetail:arrayLods areaOfInterest:extent];

    [SVProgressHUD showWithStatus:@"Estimating\n size" maskType:SVProgressHUDMaskTypeGradient];
    [self.tileCacheTask estimateTileCacheSizeWithParameters:params status:^(AGSResumableTaskJobStatus status, NSDictionary *userInfo) {
        NSLog(@"%@, %@", AGSResumableTaskJobStatusAsString(status) ,userInfo);
    } completion:^(AGSExportTileCacheSizeEstimate *tileCacheSizeEstimate, NSError *error) {
        if ( error != nil) {
            [[[UIAlertView alloc]initWithTitle:@"Error" message:[error localizedDescription] delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil] show];
            [SVProgressHUD dismiss];
        }else{
            NSNumberFormatter* tileCountFormatter = [[NSNumberFormatter alloc]init];
            [tileCountFormatter setNumberStyle:NSNumberFormatterDecimalStyle];
            [tileCountFormatter setMaximumFractionDigits:0];
            
            NSString* tileCountString = [tileCountFormatter stringFromNumber:[NSNumber numberWithInteger:tileCacheSizeEstimate.tileCount]];
            
            NSByteCountFormatter* byteCountFormatter = [[NSByteCountFormatter alloc]init];
            NSString* byteCountString = [byteCountFormatter stringFromByteCount:tileCacheSizeEstimate.fileSize];
            self.estimateLabel.text = [[NSString alloc] initWithFormat:@"%@  / %@ tiles", byteCountString, tileCountString];
            [SVProgressHUD showSuccessWithStatus:[[NSString alloc] initWithFormat:@"%@  / %@ tiles", byteCountString, tileCountString]];
        
        }

    }];
    
}



- (IBAction)downloadAction:(id)sender
{
    
    NSArray *arrayLods = [self generateLods];
    NSLog(@"LODs to be requested for the cache : %@", arrayLods);
    
    // Get the map coordinate extent from view control
    AGSEnvelope *extent = [self.mapView visibleAreaEnvelope];
    
    AGSExportTileCacheParams *params = [[AGSExportTileCacheParams alloc] initWithLevelsOfDetail:arrayLods areaOfInterest:extent];

    self.estimateLabel.text = @"";
    [SVProgressHUD showWithStatus:@"Preparing\n to download" maskType:SVProgressHUDMaskTypeGradient];
    
    self.operationToCancel = [self.tileCacheTask exportTileCacheWithParameters:params downloadFolderPath:nil useExisting:YES status:^(AGSResumableTaskJobStatus status, NSDictionary *userInfo) {
          NSLog(@"%@, %@", AGSResumableTaskJobStatusAsString(status) ,userInfo);
        NSArray *allMessages =  [userInfo objectForKey:@"messages"];
        
        if (status == AGSResumableTaskJobStatusFetchingResult) {
            NSNumber* totalBytesDownloaded = userInfo[@"AGSDownloadProgressTotalBytesDownloaded"];
            NSNumber* totalBytesExpected = userInfo[@"AGSDownloadProgressTotalBytesExpected"];
            if(totalBytesDownloaded!=nil && totalBytesExpected!=nil){
                double dPercentage = (double)([totalBytesDownloaded doubleValue]/[totalBytesExpected doubleValue]);
                [SVProgressHUD showProgress:dPercentage status:@"Downloading" maskType:SVProgressHUDMaskTypeGradient];
            }
        }else if ( allMessages.count) {
            [SVProgressHUD showWithStatus:[MessageHelper extractMostRecentMessage:allMessages] maskType:SVProgressHUDMaskTypeGradient];
        }
        
        
    } completion:^(AGSLocalTiledLayer *localTiledLayer, NSError *error) {
        [SVProgressHUD dismiss];
        if (error){
            [[[UIAlertView alloc]initWithTitle:@"Error" message:[error localizedDescription] delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil] show];
            self.estimateLabel.text = @"";
        }
        else{
            [self.mapView reset];
            [self.mapView addMapLayer:localTiledLayer withName:@"offline"];
            [self.mapView zoomToEnvelope:localTiledLayer.fullEnvelope animated:NO];
            
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Download complete" message:@"The tile cache has been added to the map." delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil];
            [alert show];
            
            [[self.floatingView subviews]
             makeObjectsPerformSelector:@selector(removeFromSuperview)];
            
            [BackgroundHelper postLocalNotificationIfAppNotActive:@"Tile cache downloaded."];
        }
    }];
}



- (NSArray*) generateLods
{
    
    AGSLOD* displayedLOD  = self.tiledLayer.currentLOD;
    NSInteger index = [self.tiledLayer.mapServiceInfo.tileInfo.lods indexOfObject:displayedLOD];
    NSRange range = NSMakeRange( index, self.scaleStepper.value );
    NSArray *lods = [self.tiledLayer.mapServiceInfo.tileInfo.lods subarrayWithRange: range];
    NSMutableArray *arrayLods = [[NSMutableArray alloc] init];
    for (AGSLOD* lod  in lods) {
        [arrayLods addObject:[NSNumber numberWithInteger:lod.level]];
    }
    
    return arrayLods;
    

}


@end

File: GpsSample/swift/Gps/Gps/ViewController.swift
//
// Copyright 2014 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//


import UIKit
import ArcGIS

class ViewController: UIViewController {
    
    
    @IBOutlet weak var label:UILabel!
    @IBOutlet weak var northArrowImage:UIImageView!
    @IBOutlet weak var autoPanModeControl:UISegmentedControl!
    @IBOutlet weak var mapView:AGSMapView!
    
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let mapUrl = NSURL(string: "http://services.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer")
        let tiledLyr = AGSTiledMapServiceLayer(URL: mapUrl)
        self.mapView.addMapLayer(tiledLyr, withName:"Tiled Layer")
        
        //Listen to KVO notifications for map gps's autoPanMode property
        self.mapView.locationDisplay.addObserver(self, forKeyPath: "autoPanMode", options: .New, context: nil)
        
        //Listen to KVO notifications for map rotationAngle property
        self.mapView.addObserver(self, forKeyPath: "rotationAngle", options: .New, context: nil)
        
        //to display actual images in iOS 7 for segmented control
        let index = self.autoPanModeControl.numberOfSegments
        for var i = 0; i < index; i++ {
            if let image = self.autoPanModeControl.imageForSegmentAtIndex(i) {
                let newImage = image.imageWithRenderingMode(.AlwaysOriginal)
                self.autoPanModeControl.setImage(newImage, forSegmentAtIndex:i)
            }
        }
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        
        //if autoPanMode changed
        if keyPath == "autoPanMode" {
            //Update the label to reflect which autoPanMode is active
            var mode:String!
            switch (self.mapView.locationDisplay.autoPanMode) {
            case .Off:
                mode = "Off"
            case .Default:
                mode = "Default"
            case .Navigation:
                mode = "Navigation"
            case .CompassNavigation:
                mode = "Compass Navigation"
            }
            self.label.textColor = UIColor.whiteColor()
            self.label.text = "AutoPan Mode: \(mode)"
            
            //Un-select the segments when autoPanMode changes to OFF
            //Also, restore north-up map rotation
            if self.mapView.locationDisplay.autoPanMode == .Off {
                self.autoPanModeControl.selectedSegmentIndex = -1
            }
            
            //Also, restore north-up map rotation if Auto pan goes OFF or back to Default
            if(self.mapView.locationDisplay.autoPanMode == .Off || self.mapView.locationDisplay.autoPanMode == .Default){
                self.mapView.setRotationAngle(0, animated:true)
                self.northArrowImage.transform = CGAffineTransformIdentity
            }
        }
            //if rotationAngle changed
        else if keyPath == "rotationAngle" {
            if self.mapView.locationDisplay.autoPanMode != .Off || self.mapView.locationDisplay.autoPanMode != .Default {
                let angle = -(self.mapView.rotationAngle*3.14)/180
                let transform = CGAffineTransformMakeRotation(CGFloat(angle))
                self.northArrowImage.transform = transform
            }
        }
            
            //if mapscale changed
        else if keyPath == "mapScale" {
            if self.mapView.mapScale < 5000 {
                self.mapView.zoomToScale(50000, withCenterPoint:nil, animated:true)
                self.mapView.removeObserver(self, forKeyPath:"mapScale")
            }
        }
    }
    
    //MARK: - Action methods
    
    @IBAction func autoPanModeChanged(sender:UISegmentedControl) {
        //Start the map's gps if it isn't enabled already
        if !self.mapView.locationDisplay.dataSourceStarted {
            self.mapView.locationDisplay.startDataSource()
        }
        
        //Listen to KVO notifications for map scale property
        self.mapView.addObserver(self, forKeyPath: "mapScale", options: .New, context: nil)
        
        //Set the appropriate AutoPan mode
        switch self.autoPanModeControl.selectedSegmentIndex {
        case 0:
            self.mapView.locationDisplay.autoPanMode = .Default
            //Set a wander extent equal to 75% of the map's envelope
            //The map will re-center on the location symbol only when
            //the symbol moves out of the wander extent
            self.mapView.locationDisplay.wanderExtentFactor = 0.75
        case 1:
            self.mapView.locationDisplay.autoPanMode = .Navigation
            //Position the location symbol near the bottom of the map
            //A value of 1 positions it at the top edge, and 0 at bottom edge
            self.mapView.locationDisplay.navigationPointHeightFactor = 0.15
        case 2:
            self.mapView.locationDisplay.autoPanMode = .CompassNavigation
            //Position the location symbol in the center of the map
            self.mapView.locationDisplay.navigationPointHeightFactor = 0.5
            
        default:
            break
        }
    }
}

